go/vt/vttablet/tabletserver/vstreamer/helper_event_test.go

```diff
@@ -0,0 +1,569 @@
+	metadata        map[string][]string        // list of enum/set values for enum/set columns
+}
+
+func (ts *TestSpec) getCurrentState(table string) *query.Row {
+	return ts.state[table]
+}
+
+func (ts *TestSpec) setCurrentState(table string, row *query.Row) {
+	ts.state[table] = row
+}
+
+// Init() initializes the test. It creates the tables and sets up the internal state.
+func (ts *TestSpec) Init() error {
+	var err error
+	if ts.inited {
+		return nil
+	}
+
+	ts.inited = true
+	if ts.options == nil {
+		ts.options = &TestSpecOptions{}
+	}
+	ts.schema, err = schemadiff.NewSchemaFromQueries(schemadiff.NewTestEnv(), ts.ddls)
+	if err != nil {
+		return err
+	}
+	ts.fieldEvents = make(map[string]*TestFieldEvent)
+	ts.fieldEventsSent = make(map[string]bool)
+	ts.state = make(map[string]*query.Row)
+	ts.metadata = make(map[string][]string)
+	ts.pkColumns = make(map[string][]string)
+	// create tables
+	for i, t := range ts.schema.Tables() {
+		execStatement(ts.t, ts.ddls[i])
+		fe := ts.getFieldEvent(t)
+		ts.fieldEvents[t.Name()] = fe
+
+		var pkColumns []string
+		var hasPK bool
+		for _, index := range t.TableSpec.Indexes {
+			if index.Info.Type == sqlparser.IndexTypePrimary {
+				for _, col := range index.Columns {
+					pkColumns = append(pkColumns, col.Column.String())
+				}
+				hasPK = true
+			}
+		}
+		if !hasPK {
+			// add all columns as pk columns
+			for _, col := range t.TableSpec.Columns {
+				pkColumns = append(pkColumns, col.Name.String())
+			}
+		}
+		ts.pkColumns[t.Name()] = pkColumns
+	}
+	engine.se.Reload(context.Background())
+	return nil
+}
+
+// Close() should be called (via defer) at the end of the test to clean up the tables created in the test.
+func (ts *TestSpec) Close() {
+	dropStatement := fmt.Sprintf("drop tables %s", strings.Join(ts.schema.TableNames(), ", "))
+	execStatement(ts.t, dropStatement)
+}
+
+func (ts *TestSpec) getBindVarsForInsert(stmt sqlparser.Statement) (string, map[string]string) {
+	bv := make(map[string]string)
+	ins := stmt.(*sqlparser.Insert)
+	tn, _ := ins.Table.TableName()
+	table := tn.Name.String()
+	fe := ts.fieldEvents[table]
+	vals, _ := ins.Rows.(sqlparser.Values)
+	for _, val := range vals {
+		for i, v := range val {
+			bufV := sqlparser.NewTrackedBuffer(nil)
+			v.Format(bufV)
+			s := bufV.String()
+			switch fe.cols[i].dataTypeLowered {
+			case "varchar", "char", "binary", "varbinary", "blob", "text":
+				s = strings.Trim(s, "'")
+			case "set", "enum":
+				s = ts.getMetadataMap(table, fe.cols[i], s)
+			}
+			bv[fe.cols[i].name] = s
+		}
+	}
+	return table, bv
+}
+
+func (ts *TestSpec) getBindVarsForUpdate(stmt sqlparser.Statement) (string, map[string]string) {
+	bv := make(map[string]string)
+	upd := stmt.(*sqlparser.Update)
+	buf := sqlparser.NewTrackedBuffer(nil)
+	upd.TableExprs[0].(*sqlparser.AliasedTableExpr).Expr.Format(buf)
+	table := buf.String()
```

Comment:
we can use `sqlparser.String(buf)` instead so that it's properly escaped if needed.